### DISCUSSION TITLE
Crazyflie 2.1 IMU Data

### DISCUSSION THREAD

**Discussion by GNC-JY** [bitcraze/discussions#171] [Q&A] [answered]
Hello.
I'm master degree student of Sejong Univ of South Korea.

Recently, I'm researching on attitude control in Harsh Condition by throwing Crazyflie.
I'm going to study so that Crazyflie can attitude control even in an inverted state.
When I checked the firmware, it seems that only Euler Angle comes out of the data that passed EKF.
it is impossible to calculate data when it reaches 90 degrees or -90 degrees. So we need the quaternion data.
Can I get the quaternion data from Crazyflie's EKF?

**Comment by krichardsson** [comment, marked as answer]
Hi!

The EKF is using quaternions internally and the quaternion data is available as log data in stateEstimate.qx, stateEstimate.qy, stateEstimate.qz, stateEstimate.qw
See https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/stabilizer.c#L701-L719

It is also available as a compressed quaternion in stateEstimateZ.quat
See https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/stabilizer.c#L776-L779

**Reply by GNC-JY** [reply]
So what does q0 to q3 in estimator_kalman.c mean?
I can't find There's nothing that says where they're coming from.
We want to get the data of state.attitudeQuaternion.x,y,z,w.

**Reply by krichardsson** [reply]
[q0-q3](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/estimator_kalman.c#L552-L566) contains the same values, but from a different part of the memory, slightly earlier in the process. The values in the [kalman core data](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/estimator_kalman.c#L134) are first [externalized in the kalman task](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/estimator_kalman.c#L290), and finally [copied to the stabilizer memory](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/estimator_kalman.c#L304).

You could pretty much use any one of them since the contain the same data. 
If you are writing code in the Crazyflie firmware, you should use the flavor that is easiest to use from where your code is running.

**Reply by GNC-JY** [reply]
I appreciate your reply.

**Reply by knmcguire** [reply]
@GNC-JY if this answer has satisfied your initial request, could you mark it as answer than by pressing 'mark as answer' ?
